dotCMS/src/main/java/com/dotcms/auth/providers/saml/v1/SAMLHelper.java

```diff
@@ -0,0 +1,469 @@
+        Logger.debug(this, ()->"Default SAML User role has been assigned");
+
+        // the only strategy that does not include the saml user role is the "idp"
+        if (!DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_IDP_VALUE.equalsIgnoreCase(buildRolesStrategy)) {
+            // Add DOTCMS_SAML_OPTIONAL_USER_ROLE
+            if (DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                    SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE) != null) {
+
+                this.addRole(user, DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                        SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE), false, false);
+                Logger.debug(this, ()-> "Optional user role: " +
+                        DotSamlProxyFactory.getInstance().samlConfigurationService().getConfigAsString(identityProviderConfiguration,
+                                SamlName.DOTCMS_SAML_OPTIONAL_USER_ROLE) + " has been assigned");
+            }
+        } else {
+
+            Logger.info(this, "The build roles strategy is 'idp'. No saml_user_role has been added");
+        }
+    }
+
+    private boolean isValidRole(final String role, final String[] rolePatterns) {
+
+        boolean isValidRole = false;
+
+        if (null != rolePatterns) {
+            for (final String rolePattern : rolePatterns) {
+                Logger.debug(this, ()-> "Valid Role: " + role + ", pattern: " + rolePattern);
+                isValidRole |= this.match(role, rolePattern);
+            }
+        } else {
+            // if not pattern, role is valid.
+            isValidRole = true;
+        }
+
+        return isValidRole;
+    }
+
+    private boolean match(final String role, final String rolePattern) {
+        String uftRole = null;
+
+        try {
+
+            uftRole = URLDecoder.decode(role, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            uftRole = role;
+        }
+
+        return RegEX.contains(uftRole, rolePattern);
+    }
+
+    private void addRolesFromIDP(final User user, final Attributes attributesBean, final IdentityProviderConfiguration identityProviderConfiguration,
+                                 final String buildRolesStrategy) throws DotDataException {
+
+        final boolean includeIDPRoles = DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_ALL_VALUE.equalsIgnoreCase(buildRolesStrategy)
+                || DotSamlConstants.DOTCMS_SAML_BUILD_ROLES_IDP_VALUE.equalsIgnoreCase(buildRolesStrategy);
+
+        Logger.debug(this, ()-> "Including roles from IdP '" + includeIDPRoles + "' for the build roles Strategy: " + buildRolesStrategy);
+
+        if (includeIDPRoles && attributesBean.isAddRoles() && null != attributesBean.getRoles()) {
+
+            final List<String> roleList = this.samlAuthenticationService.getValues(attributesBean.getRoles());
+            if (null != roleList && roleList.size() > 0) {
```

Comment:
Codacy found an issue: [Substitute calls to size() == 0 (or size() != 0, size() > 0, size() < 1) with calls to isEmpty()](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5707914)